waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/tunnelling/TunnelingMetaStoreClientFactory.java

```diff
@@ -41,55 +38,65 @@
   @VisibleForTesting
   final MethodChecker METHOD_CHECKER = new MetastoreClientMethodChecker();
 
-  private final TunnelableFactorySupplier tunnelableFactorySupplier;
+  private TunnelableFactory<CloseableThriftHiveMetastoreIface> tunnelableFactory;
+  private String localhost;
+  private HiveConf localHiveConf;
   private final MetaStoreClientFactory defaultFactory;
-  private final LocalHiveConfFactory localHiveConfFactory;
   private final HiveMetaStoreClientSupplierFactory hiveMetaStoreClientSupplierFactory;
 
   public TunnelingMetaStoreClientFactory() {
-    this(new TunnelableFactorySupplier(), new DefaultMetaStoreClientFactory(), new LocalHiveConfFactory(),
-        new HiveMetaStoreClientSupplierFactory());
+    this(new DefaultMetaStoreClientFactory(), new HiveMetaStoreClientSupplierFactory());
   }
 
   @VisibleForTesting
   TunnelingMetaStoreClientFactory(
-      TunnelableFactorySupplier tunnelableFactorySupplier,
       MetaStoreClientFactory defaultFactory,
-      LocalHiveConfFactory localHiveConfFactory,
       HiveMetaStoreClientSupplierFactory hiveMetaStoreClientSupplierFactory) {
-    this.tunnelableFactorySupplier = tunnelableFactorySupplier;
     this.defaultFactory = defaultFactory;
-    this.localHiveConfFactory = localHiveConfFactory;
     this.hiveMetaStoreClientSupplierFactory = hiveMetaStoreClientSupplierFactory;
   }
 
+  @Override
+  public void setTunnelableFactory(TunnelableFactory<CloseableThriftHiveMetastoreIface> tunnelableFactory) {
```

Comment:
So before these changes the constructor took this and localHost as paramaters and once you had constructed the factory then calles to `newInstance()` would always return the same type. Now theoretically you could construct the factory without setting this or the next one, make some calls to `newInstance()` and get a certain type of `CloseableThriftHiveMetastoreIface` back, then set these, call `newInstance()` again and get a different (tunnelled) type back. I see from the flow of code in the classes above that this wouldn't actually happen in how we use this factory but it is a bit unusual. How hard is it to get these two handed over via a constructor?

Reply:
I don't think it should be hard to do it, but I'm going to try it out and update you on this.